### PR TITLE
Fix mounting boot device on CentOS 7.6

### PR DESCRIPTION
**Pull request by bgilbert**
The CentOS 7.6 `lsblk` doesn't support `--json` or the PATH key.  Switch to `--pairs`, and form the device path from `NAME`.

### CODE REVIEW

File: src/blockdev.rs [hunk 1]
```diff
@@ -61,17 +56,48 @@ fn get_partition_with_label(device: &str, label: &str) -> Result<Option<BlkDev>>
...
+        let mut path = Path::new(device)
+            .parent()
+            .chain_err(|| format!("couldn't get parent directory of {}", device))?
+            .to_path_buf();
+        match fields.get("NAME") {
+            None => continue,
+            Some(name) => path.push(name),
+        }
+
+        // accept
```
Comment (lucab) [inline src/blockdev.rs]: Spurious line?

File: src/blockdev.rs [hunk 2]
```diff
@@ -61,17 +56,48 @@ fn get_partition_with_label(device: &str, label: &str) -> Result<Option<BlkDev>>
         eprint!("{}", String::from_utf8_lossy(&*result.stderr));
         bail!("lsblk of {} failed", device);
     }
-    let output: LsBlk =
-        serde_json::from_slice(&*result.stdout).chain_err(|| "decoding lsblk JSON")?;
+    let output = String::from_utf8(result.stdout).chain_err(|| "decoding lsblk output")?;
+
+    // walk each device in the output
+    let re = Regex::new("([A-Z-]+)=\"([^\"]+)\"").unwrap();
```
Comment (lucab) [inline src/blockdev.rs]: You can use raw strings here if you prefer avoiding quote-escapes: `r#"a "raw" string"#`

File: src/blockdev.rs [hunk 3]
```diff
@@ -61,17 +56,48 @@ fn get_partition_with_label(device: &str, label: &str) -> Result<Option<BlkDev>>
...
     let mut found: Option<BlkDev> = None;
-    for dev in output.blockdevices {
-        if dev.label.is_none() || dev.label.as_ref().unwrap() != label {
-            continue;
+    for line in output.lines() {
+        // parse key-value pairs
+        // newer versions of lsblk support JSON but CentOS 7 doesn't
+        let mut fields: HashMap<String, String> = HashMap::new();
+        for cap in re.captures_iter(line) {
+            fields.insert(cap[1].to_string(), cap[2].to_string());
```
Comment (lucab) [inline src/blockdev.rs]: Should we check-and-continue for `cap.len() < 2` here?

File: src/blockdev.rs [hunk 4]
```diff
@@ -61,17 +56,48 @@ fn get_partition_with_label(device: &str, label: &str) -> Result<Option<BlkDev>>
...
+    // walk each device in the output
+    let re = Regex::new("([A-Z-]+)=\"([^\"]+)\"").unwrap();
     let mut found: Option<BlkDev> = None;
-    for dev in output.blockdevices {
-        if dev.label.is_none() || dev.label.as_ref().unwrap() != label {
-            continue;
+    for line in output.lines() {
+        // parse key-value pairs
+        // newer versions of lsblk support JSON but CentOS 7 doesn't
+        let mut fields: HashMap<String, String> = HashMap::new();
```
Comment (lucab) [inline src/blockdev.rs]: If you have time for it here, I would move the output parsing logic to its own function, returning this hashmap.
That way we can have a test covering it with some real-world `lsblk` output, in turn making it easier to see what's going on.

File: src/blockdev.rs (hunk 1 above)
Comment (bgilbert) [reply src/blockdev.rs]: Uh, no?

File: src/blockdev.rs (hunk 3 above)
Comment (bgilbert) [reply src/blockdev.rs]: Neither of the match groups in the regex are optional, so we shouldn't end up here if `cap.len() < 3`.  I'm okay with the panic in that case.

File: src/blockdev.rs (hunk 1 above)
Comment (lucab) [reply src/blockdev.rs]: Does it spark joy?

On a serious note: :+1:, sorry, it was hanging a bit alone so my brain put it in the garbage pile.

**Comment by bgilbert** [comment]
Updated.

**Pull request merged**